### DISCUSSION TITLE
Why is the custom error handler ignored?

### DISCUSSION THREAD

**Discussion by H3xept** [fastify/fastify#5257] [Q&A] [answered]
Hey all. 

I noticed strange behavior when using AJV and custom error handlers. It seems like if I register a custom error handler **after** routes are registered, AJV validation errors won't trigger the custom handler. Setting the error handler before route registration works.

Here's a minimal example:

```typescript
const fastify = require('fastify')({
  logger: true
})

const schema = {
    schema: {
        querystring: {
            type: 'object',
            required: [
                'test'
            ]
        }
    }
}

const errorHandler = () => {
    throw new Error(`Custom error`)
}

async function registerRoutes(
    app,
    _,
    next
)  {
    app.get('/', schema , function (reply) {
        reply.send({ hello: 'world' })
    }, )
    next()
}

fastify.register(registerRoutes).then(() => {
    fastify.setErrorHandler(errorHandler)
    fastify.listen({ port: 3000 }, function (err, address) {
        if (err) {
            fastify.log.error(err)
            process.exit(1)
        }
    })
})
```

Expected response for `curl http://127.0.0.1:3000 ` : 
![image](https://github.com/fastify/fastify/assets/2883651/fe5b4381-4688-45b5-b3ae-57d389875129)

Actual response:
![image](https://github.com/fastify/fastify/assets/2883651/a27420e1-e142-447f-88c5-ba335019877c)

Why does this happen? Is this intended behavior?

**Comment by Eomm** [comment, marked as answer]
if you use `then`, the plugin is loaded and you can't set the errorHandler after initializing it.

```diff
+fastify.setErrorHandler(errorHandler)
fastify.register(registerRoutes).then(() => {
-    fastify.setErrorHandler(errorHandler)
    fastify.listen({ port: 3000 }, function (err, address) {
        if (err) {
            fastify.log.error(err)
            process.exit(1)
        }
    })
})
```


or

```
fastify.register(registerRoutes)
fastify.setErrorHandler(errorHandler)
fastify.listen({ port: 3000 }, function (err, address) {
    if (err) {
        fastify.log.error(err)
        process.exit(1)
    }
})
```

**Reply by H3xept** [reply]
Ok so this is intended behavior. Thanks for the clarification.

**Comment by cesarvspr** [comment]
I was having same error - fixed by registering error handler before any route.
